<a href="https://colab.research.google.com/github/opoluliakh/data-analytics-learning/blob/main/Poluliakh_hw_3_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/supermarket_sales.csv')
type(df)

Mounted at /content/drive


pandas.core.frame.DataFrame

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [2]:

branch_profit = df.groupby('Branch')['Total'].sum().sort_values(ascending=False)


print(branch_profit)

# Найприбутковішим філіалом є філіал C, адже він має найвищу загальну суму прибутку серед усіх трьох філіалів.


Branch
C    110568.7065
A    106200.3705
B    106197.6720
Name: Total, dtype: float64


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [3]:
city_branch_profit = df.groupby(['City', 'Branch'])['Total'].sum().sort_values(ascending=False)
print(city_branch_profit.head(1))


City       Branch
Naypyitaw  C         110568.7065
Name: Total, dtype: float64


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [4]:
invoices_by_category = df.pivot_table(
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

print(invoices_by_category)


Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45  


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [5]:
top_branch = invoices_by_category['Electronic accessories'].idxmax()
print(top_branch)


A


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Питання 1: У який день тижня здійснюють найбільше покупок?**

Гіпотеза. Покупці найактивніші у певні дні тижня, наприклад, у вихідні.

In [6]:
df['Date'] = pd.to_datetime(df['Date'])
df['Weekday'] = df['Date'].dt.day_name()
weekday_counts = df.groupby('Weekday')['Invoice ID'].count().sort_values(ascending=False)
print(weekday_counts)


Weekday
Saturday     164
Tuesday      158
Wednesday    143
Friday       139
Thursday     138
Sunday       133
Monday       125
Name: Invoice ID, dtype: int64


**Висновок**: Найбільше покупок здійснюється в суботу, а найменше — в понеділок. Це частково підтверджує гіпотезу, що у вихідні активність покупців зростає. Субота — найактивніший день, тоді як неділя виявилася менш активною, ніж будні дні, такі як вівторок і середа. Це може свідчити, що частина клієнтів відкладає покупки на кінець робочого тижня або планує їх на суботу.

Рекомендації: Зосередити рекламні кампанії або знижки на днях із меншими продажами, аби вирівняти навантаження. Оптимізувати графік роботи та персонал на суботу. Проводити акції у понеділок для стимулювання продажів у найменш активний день

**Питання 2: Яка категорія товарів приносить найбільше прибутку?**

Гіпотеза: Деякі категорії стабільно приносять більше грошей.


In [8]:
category_revenue = df.groupby('Product line')['Total'].sum().sort_values(ascending=False)
print(category_revenue)


Product line
Food and beverages        56144.8440
Sports and travel         55122.8265
Electronic accessories    54337.5315
Fashion accessories       54305.8950
Home and lifestyle        53861.9130
Health and beauty         49193.7390
Name: Total, dtype: float64


**Висновок**: Результати показують, що найбільше прибутку приносять товари з категорії "Food and beverages", які згенерували сумарно понад 56 тисяч доларів. Далі йдуть "Sports and travel" (близько 55 000) та "Electronic accessories" (біля 54 000). Найменше прибутку принесли товари з категорії "Health and beauty" (трохи більше 49 000).

Це підтверджує гіпотезу, що певні категорії товарів стабільно приносять більше грошей, і може вказувати на вищу популярність або вартість продуктів у цих категоріях. Бізнес може розглянути можливість розширення асортименту саме в цих напрямах.

Рекомендація: варто зосередити маркетингові зусилля та розширити асортимент у категоріях "Food and beverages", "Sports and travel" та "Electronic accessories", оскільки саме вони генерують найбільший прибуток.


**Питання 3: Чи існує відмінність у середньому чеку (Total) між чоловіками та жінками?**

Гіпотеза: стать впливає на середній чек. Це дозволяє зрозуміти споживчу поведінку різних груп клієнтів.

In [9]:
df.groupby('Gender')['Total'].mean().sort_values(ascending=False)


Gender
Female    335.095659
Male      310.789226
Name: Total, dtype: float64

А тепер давайте подивимось по категоріях.

In [10]:
gender_category_avg = df.pivot_table(
    values='Total',
    index='Product line',
    columns='Gender',
    aggfunc='mean'
).sort_values(by='Female', ascending=False)

print(gender_category_avg)


Gender                      Female        Male
Product line                                  
Home and lifestyle      380.213639  294.136241
Food and beverages      368.565750  273.499125
Sports and travel       324.712739  340.360327
Electronic accessories  322.643125  316.691965
Fashion accessories     317.056250  291.079207
Health and beauty       290.015414  348.099460


**Висновок:** Аналіз середнього чека показав, що жінки в середньому витрачають більше, ніж чоловіки (335 доларів проти 311 доларів). Однак при розбивці за категоріями виявилось, що це не універсальна тенденція: жінки витрачають більше на товари для дому, їжу, моду, а чоловіки – на подорожі, електроніку та особисту гігієну. Отже, стать справді впливає на споживчі вподобання, і середній чек варто аналізувати в контексті категорій товарів.

Рекомендація: адаптувати маркетингові стратегії з урахуванням гендерних вподобань. Наприклад, рекламувати категорії «Home and lifestyle» та «Food and beverages» для жінок, а «Health and beauty» або «Sports and travel» – для чоловіків. Це дозволить підвищити релевантність реклами та стимулювати витрати в кожному сегменті.